### ISSUE TITLE
Classifier defaults the score depends on

### ISSUE THREAD

**Issue by DEVunderdog** [DEVunderdog/DataForgeML#493]
Part of #486

## Question

Which of the classifier's defaults does the library own, and which does it inherit from sklearn?

[Does the C2ST null actually hold on our construction?](https://github.com/DEVunderdog/DataForgeML/issues/487) found that `HistGradientBoostingClassifier`'s stock defaults silently break the test at small pile sizes. `min_samples_leaf=20` cannot split fewer than 40 training rows, so the model constant-predicts, a balanced test set pins accuracy at exactly 0.5, and z comes back **0.000 — indistinguishable from a perfect imputation**. A constant fill at 30 rows per pile is missed 100% of the time. Lowering it to 5 catches the same case 98% of the time.

The map already settled that the classifier is **injectable**. This ticket is about what the *default* is when nobody injects one.

- **Does the library set `min_samples_leaf` at all?** Leaving sklearn's 20 imports a silent false negative. Options: a fixed lower value, a value scaled to pile size, or refusing to run below a floor and never touching the setting. State what each gains and gives up.
- **If it scales, scaled to what, and is the scaling visible?** A setting that changes with `n` makes two columns' z-scores products of different models. Whether that is acceptable, given the score is already per-column, needs saying out loud.
- **`early_stopping` is now off** (#487 overturned forcing it on). Is it left at sklearn's default — which enables it only above 10,000 rows — or explicitly pinned off, so behaviour does not flip when a table crosses that line?
- **Anything else load-bearing?** `max_iter`, `learning_rate`, `max_leaf_nodes`, `l2_regularization`. The test is about the classifier's *capacity*, so any default that caps capacity can manufacture a pass. #487 only measured the two above.
- **How does an injected classifier interact with all of this?** If a user injects a bare `RandomForestClassifier`, do the library's capacity decisions apply, or does injection mean the user owns every default? The map rejected RF as the *default*; it did not ban it.
- **Does the report say which classifier and which settings produced the number?** Two runs with different capacity are not comparable, and a z-score with no record of the model behind it cannot be reproduced.

Per the accuracy-over-speed principle, a default that quietly costs detection is worse than one that costs compute.

**Comment by DEVunderdog** [comment]
## Answer

**The library pins exactly two settings, and it pins them because sklearn's defaults encode a dataset-size assumption that balanced C2ST piles violate. Everything else is inherited. Injection is total — an injected classifier gets none of this, and the report says which classifier produced the number.**

Verified directly against this repo's pinned scikit-learn **1.9.0** before answering:

- `HistGradientBoostingClassifier` defaults: `min_samples_leaf=20`, `early_stopping='auto'`, `max_iter=100`, `learning_rate=0.1`, `max_leaf_nodes=31`, `l2_regularization=0.0`, `max_depth=None`, `max_features=1.0`, `class_weight=None`.
- `'auto'` resolves as `do_early_stopping_ = n_samples > 10_000` (`_hist_gradient_boosting/gradient_boosting.py`), where `n_samples` is the training rows of **one balanced repeat**. At the default 70/30 split that is `min(pile) > ~7,150`. The flip is reachable on a large table, not theoretical.

### 1. `min_samples_leaf` is fixed at 5, not scaled

[Columns the test cannot run on](https://github.com/DEVunderdog/DataForgeML/issues/489) already set the value and measured it (null z SD 0.92–1.04, FPR 1.5–4.5% at every pile ≥ 30, constant-predictor pathology cured). This ticket settles the half it left open: **fixed, never scaled to pile size.**

A scaled leaf floor — `max(2, min(20, m // 10))` or similar — makes two columns in one report products of different-capacity models. The map's premise is that **the z *is* the score**, the number a user reads across columns and across runs; a scaled setting makes that number a function of the column's missingness rate as well as its distributional damage. #489's table also shows scaling buys nothing: at `m ≥ 100` the two settings are indistinguishable, and below that 5 strictly dominates.

*Gained:* one model, one meaning; the score stays comparable across columns and across runs. *Given up:* on a very large pile, 5 is a lower leaf floor than HGB would have chosen, so the default classifier is more overfit-prone there — which [#487](https://github.com/DEVunderdog/DataForgeML/issues/487) measured as failing **safe** (train 1.000 against test 0.497, null FPR still 3.7–6.9%).

### 2. `early_stopping` is pinned `False`, not left at `'auto'`

[#487](https://github.com/DEVunderdog/DataForgeML/issues/487) overturned *forcing it on*. That leaves two distinct options, and inheriting `'auto'` is the wrong one.

Under `'auto'`, a table where one column's piles cross ~7,150 and another's do not gets **two different algorithms in one report**, with nothing in the output saying so. Worse, the flip cuts capacity on exactly the piles with the most power, and #487 measured that cut costing real detection (degenerate imputer: 100% → 89%).

*Gained:* one algorithm at every table size; behaviour does not flip when a table crosses a line. *Given up:* on huge piles we always run all 100 iterations, so the default is slower than sklearn's; and reports keep showing train accuracy 1.000, the cost #487 already accepted under guard 1.

### 3. Nothing else is pinned — and the rule that decides it

The ticket named `max_iter`, `learning_rate`, `max_leaf_nodes`, `l2_regularization` as capacity-capping candidates. All four were checked; none is tuned against a row count.

**The rule, which is the durable part of this decision:**

> The library pins a classifier setting only where sklearn's default is *structurally wrong for balanced C2ST piles* — that is, where the default encodes an assumption about dataset size. It does not pin settings that merely express a tuning preference.

`min_samples_leaf=20` (cannot split fewer than 40 training rows) and `early_stopping='auto'` (flips at 10,000) are the only two defaults that qualify. `max_iter=100` and `learning_rate=0.1` cap capacity only in the sense that any finite budget does; raising them is *tuning the classifier*, which is what the injection channel is for. `class_weight=None` is correct because balancing is mandatory. `max_bins=255`, `max_depth=None`, `max_features=1.0`, `l2_regularization=0.0`, `max_leaf_nodes=31` are all size-agnostic.

*Gained:* a stated, testable boundary — the next person who wants to pin `max_iter` has a test to apply rather than a taste to argue. *Given up:* it is possible some other default manufactures a pass at a pile size nobody has measured, and this rule would not have caught it; only measurement would. That is a stated limit, not a fixed one.

**This breaks a repo precedent, and the ADR must say so out loud.** `_STRATEGY_DIALS` (`imputation/_config.py:80`) states that every imputation dial value "is also the sklearn default of the estimator it reaches (`IterativeImputer.max_iter` = 10, `KNNImputer.n_neighbors` = 5, and so on)". C2ST is the first place the library overrides a sklearn default. The justification is that those dials are *decide-time formulas degenerating to neutral inputs*, whereas these two are corrections to an assumption sklearn makes about `n` that C2ST structurally violates.

### 4. The pins live in a generic-side factory — closing a gap between #489 and #490

[#489](https://github.com/DEVunderdog/DataForgeML/issues/489) made `min_samples_leaf` a **`C2STConfig` dial**. [The generic C2ST contract at the module seam](https://github.com/DEVunderdog/DataForgeML/issues/490) ruled that **`C2STConfig` never crosses the seam**, and gave the generic function `classifier: ClassifierMixin | None = None`.

Taken together, those two leave a hole: **`c2st(..., classifier=None)` builds a bare HGB with sklearn's `min_samples_leaf=20`** — the exact pathology this ticket exists to kill, reachable through the private generic door and through every future non-imputation adapter.

**Resolution: a generic-side factory owns the pinned default.**

```python
_DEFAULT_MIN_SAMPLES_LEAF = 5

def _default_c2st_classifier(
    *, min_samples_leaf: int = _DEFAULT_MIN_SAMPLES_LEAF, random_state: int | None = None
) -> ClassifierMixin: ...
```

It applies both pins (`min_samples_leaf`, `early_stopping=False`) and inherits everything else. The imputation adapter calls it with the `C2STConfig` dial; `c2st`'s `classifier=None` branch calls it with the parameter default. Same argument #490 used to put the pooling helper generic-side: the rule people get wrong gets one home and one docstring.

Rejected: **`min_samples_leaf` crossing the seam as a plain keyword** beside `repeats` and `scheme` — those two cross because they are *mechanism* (balancing's answer is unstable without `R`; the scheme changes what `n_te` means), whereas a leaf floor is a property of one particular classifier and would be meaningless the moment someone injects a different one. Rejected: **dropping the `None` default and requiring a classifier** — it pushes the pathology onto every future adapter author instead of removing it.

**One module-level constant, read from both places**, which closes the drift this option otherwise gives up: `C2STConfig.min_samples_leaf` takes `_DEFAULT_MIN_SAMPLES_LEAF` as its field default, and the factory takes it as its parameter default. There is one number, in one place.

*Gained:* the pins have exactly one home; the private generic door cannot hand out a sklearn-default HGB by accident; `C2STConfig` still never crosses the seam. *Given up:* one more generic-side symbol.

### 5. Injection is total — the user owns every default of an injected classifier

An injected instance is used **verbatim**. The library forces nothing onto it.

The alternative — reaching in with `set_params(min_samples_leaf=…)` — breaks on any classifier lacking that parameter (`RandomForestClassifier` has it; most do not), so it needs a `try`/`except` that silently does nothing on failure. That is the worst outcome available: the user cannot tell whether their setting survived. Direct repo precedent: `ImputationDecision.custom_estimators` holds the user's live instance, and [#450](https://github.com/DEVunderdog/DataForgeML/issues/450) ruled it is not even cloned.

The map rejected `RandomForestClassifier` as the **default** (cannot eat text without an encoder the library would have to own and defend; grown-to-purity is the saturation-prone end of Snoke's failure mode 2). It did not ban it, and this decision keeps it usable.

The safety net already exists and is the right shape: **#489's runtime degeneracy guard** catches a constant-predicting injected classifier and returns `Uninformative` rather than `z = 0.000`.

*Gained:* injection means what it says, and works for any sklearn classifier. *Given up, and this is a real hole:* an injected bare `HistGradientBoostingClassifier` reintroduces the pathology at 30–60 rows per pile, precisely where #489 measured degeneracy as an **unreliable** detector (finding (c): at `m=30, mls=20` the classifier is not yet constant and detection is 6%). A user who injects a bare HGB gets a worse test than the default, with only the `classifier_injected` flag in provenance (§6) to warn them. That is the price of not fighting the user's object, and it is documented rather than designed away.

### 6. Provenance records the model, and records the params sklearn's `repr` hides

[What the C2ST result record reports](https://github.com/DEVunderdog/DataForgeML/issues/488) created `C2STProvenance` at report level — one object per call, since a single evaluation uses one classifier for every column — and left its contents here.

```
C2STProvenance
    classifier_class      str                  # fully-qualified, e.g. "sklearn.ensemble._hist_gradient_boosting...HistGradientBoostingClassifier"
    classifier_params     dict[str, str]       # full get_params(), values stringified
    classifier_injected   bool
    sklearn_version       str
    config                C2STConfig           # from #488: the resolved config
```

**Full `get_params()`, not `repr(estimator)`.** sklearn's `repr` elides every parameter left at its default — so the single fact you most need when arguing with a report (that `min_samples_leaf` was 20, not 5) is exactly what `repr` hides. Values are stringified because `get_params()` can hold live nested estimators that do not serialise.

**`classifier_injected` is its own field**, not inferable from the class name: "the library's default HGB" and "a user-supplied HGB that happens to look identical" answer different questions, and §5 means only the first carries the pins.

**`sklearn_version`** because §3's rule is a claim *about a specific sklearn release*. If a future release moves `min_samples_leaf`'s default, or changes `'auto'`'s threshold, this ticket's reasoning needs re-checking — and a report from before that release must be readable as such.

*Gained:* every z is reproducible from the report alone; two runs with different capacity are visibly different rather than silently incomparable. *Given up:* provenance grows a params blob that is noise in the common case, and stringified params are not round-trippable back into a live estimator.

---

### Config surface — unchanged

No new dials. `min_samples_leaf = 5` was already `C2STConfig`'s from #489; it now also has a named constant behind it. **`early_stopping` is deliberately not a dial**, and this is the one place the full-configurability principle is not applied: it is not a threshold, it is a structural correction to a size assumption, and a user who wants different classifier internals has the injection channel — a wider, better door than a single boolean.

### For the ADR

- The §3 rule, stated as the pinning boundary.
- The `_STRATEGY_DIALS` precedent break, stated explicitly with its justification.
- The §5 hole: an injected bare HGB is worse than the default at small piles, and the degeneracy guard does not fully cover it.

### Corrections to closed tickets

None. This completes [#489](https://github.com/DEVunderdog/DataForgeML/issues/489)'s `min_samples_leaf` decision (fixed, not scaled), fills [#488](https://github.com/DEVunderdog/DataForgeML/issues/488)'s deferred `C2STProvenance` contents, and closes a gap left between #489 and [#490](https://github.com/DEVunderdog/DataForgeML/issues/490) (§4) that neither ticket owned.

Resolved — two pins, one stated rule, a generic-side factory holding them, total injection, and provenance that records what `repr` hides.
